core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/BindingSetAssignmentInliner.java

```diff
@@ -0,0 +1,78 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.query.algebra.evaluation.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.rdf4j.query.BindingSet;
+import org.eclipse.rdf4j.query.Dataset;
+import org.eclipse.rdf4j.query.algebra.AbstractQueryModelNode;
+import org.eclipse.rdf4j.query.algebra.BindingSetAssignment;
+import org.eclipse.rdf4j.query.algebra.LeftJoin;
+import org.eclipse.rdf4j.query.algebra.QueryModelNode;
+import org.eclipse.rdf4j.query.algebra.TupleExpr;
+import org.eclipse.rdf4j.query.algebra.Var;
+import org.eclipse.rdf4j.query.algebra.evaluation.QueryOptimizer;
+import org.eclipse.rdf4j.query.algebra.helpers.AbstractQueryModelVisitor;
+
+/**
+ * Optimizes a query model by inlining {@link BindingSetAssignment} values where possible.
+ *
+ * @author Jeen Broekstra
+ */
+public class BindingSetAssignmentInliner implements QueryOptimizer {
+
+	@Override
+	public void optimize(TupleExpr tupleExpr, Dataset dataset, BindingSet bindings) {
+		BindingSetAssignmentVisitor visitor = new BindingSetAssignmentVisitor();
+		tupleExpr.visit(visitor);
+	}
+
+	private static class BindingSetAssignmentVisitor extends AbstractQueryModelVisitor<RuntimeException> {
+
+		private BindingSet bindingSet;
+
+		@Override
+		public void meet(BindingSetAssignment bsa) {
+			List<BindingSet> bindingSets = new ArrayList<>();
+			bsa.getBindingSets().forEach(bindingSets::add);
+
+			if (bindingSets.size() == 1) {
+				bindingSet = bindingSets.get(0);
```

Comment:
Can't you just iterate over the bindingsets, cache the first one and then check if there are anymore? Instead of creating an array list every time.

Reply:
Yep, good catch, done.